tests/bsim/bluetooth/host/central/src/main.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "bstests.h"
+#include "babblekit/testcase.h"
+#include <zephyr/bluetooth/conn.h>
+
+static K_SEM_DEFINE(sem_connected, 0, 1);
+
+static void connected_cb(struct bt_conn *conn, uint8_t err)
+{
+	TEST_ASSERT(conn);
+	TEST_ASSERT(err == BT_HCI_ERR_UNKNOWN_CONN_ID, "Expected connection timeout");
+
+	k_sem_give(&sem_connected);
+	bt_conn_unref(conn);
+}
+
+static struct bt_conn_cb conn_cb = {
+	.connected = connected_cb,
+};
+
+static void test_central_connect_timeout_with_timeout(uint16_t timeout_param,
+						      uint32_t expected_conn_timeout_ms)
+{
+	int err;
+	struct bt_conn *conn;
+
+	bt_addr_le_t peer = {.a.val = {0x01}};
+	const struct bt_conn_le_create_param create_param = {
+		.options = BT_CONN_LE_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+		.interval_coded = 0,
+		.window_coded = 0,
+		.timeout = timeout_param,
+	};
+
+	k_sem_reset(&sem_connected);
+
+	const uint64_t conn_create_start = k_uptime_get();
+
+	err = bt_conn_le_create(&peer, &create_param, BT_LE_CONN_PARAM_DEFAULT, &conn);
+	TEST_ASSERT(err == 0, "Failed starting initiator (err %d)", err);
+
+	err = k_sem_take(&sem_connected, K_SECONDS(2 * CONFIG_BT_CREATE_CONN_TIMEOUT));
+	TEST_ASSERT(err == 0, "Failed getting connected timeout within %d s (err %d)",
+		    2 * CONFIG_BT_CREATE_CONN_TIMEOUT, err);
```

Comment:
Correct me if I'm wrong, but we have this code in the host
```c
	bt_dev.create_param.timeout =
		(bt_dev.create_param.timeout != 0) ?
		bt_dev.create_param.timeout :
		(MSEC_PER_SEC / 10) * CONFIG_BT_CREATE_CONN_TIMEOUT;
```

Does that mean if `create_param.timeout` is set, then `CONFIG_BT_CREATE_CONN_TIMEOUT` isn't used?

This function seems to set both `create_param.timeout` and use `CONFIG_BT_CREATE_CONN_TIMEOUT` which doesn't look right to me

Reply:
I was also a bit confused by the code, so I added this test to convince myself that it was actually working.
See the `bt_dev.create_param = *param;` above :).

I will update the semaphore timeout here to match the expected timeout

Reply:
`CONFIG_BT_CREATE_CONN_TIMEOUT` should probably be called `CONFIG_BT_CREATE_CONN_TIMEOUT_DEFAULT` instead or maybe not exist at all. It was added before the `create_param.timeout` parameter existed and was at the time the only way to set the connection timeout.

Reply:
That would be an API breaker, so we should probably do that in a separate PR